src/pool-cl/anti-sniping/CLAntiSniping.sol

```diff
@@ -0,0 +1,249 @@
+                afterAddLiquidiyReturnsDelta: false,
+                afterRemoveLiquidiyReturnsDelta: true
+            })
+        );
+    }
+
+    /// @notice Collects fee information for positions created in the last processed block.
+    /// @dev This is called in all of the before hooks (except init) and can also be called manually.
+    /// @param poolId The identifier of the pool.
+    function collectLastBlockInfo(PoolId poolId) public {
+        if (block.number <= lastProcessedBlockNumber[poolId]) {
+            return;
+        }
+        lastProcessedBlockNumber[poolId] = block.number;
+        for (uint256 i = 0; i < positionsCreatedInLastBlock[poolId].length; i++) {
+            bytes32 positionKey = positionsCreatedInLastBlock[poolId][i];
+            LiquidityParams memory params = positionKeyToLiquidityParams[positionKey];
+            CLPosition.Info memory info = poolManager.getPosition(poolId, params.sender, params.tickLower, params.tickUpper, params.salt);
+            (uint256 feeGrowthInside0X128, uint256 feeGrowthInside1X128) = _getFeeGrowthInside(poolId, params.sender, params.tickLower, params.tickUpper, params.salt);
+            firstBlockFeesToken0[poolId][positionKey] =
+                                FullMath.mulDiv(feeGrowthInside0X128 - info.feeGrowthInside0LastX128, info.liquidity, FixedPoint128.Q128);
+            firstBlockFeesToken1[poolId][positionKey] =
+                                FullMath.mulDiv(feeGrowthInside1X128 - info.feeGrowthInside1LastX128, info.liquidity, FixedPoint128.Q128);
+        }
+        delete positionsCreatedInLastBlock[poolId];
+    }
+
+    /// @notice Handles logic after removing liquidity, redistributing first-block fees if applicable.
+    /// @dev Donates first-block accrued fees to the pool if liquidity remains; otherwise, returns them to the sender.
+    function afterRemoveLiquidity(
+        address sender,
+        PoolKey calldata key,
+        ICLPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta,
+        BalanceDelta,
+        bytes calldata
+    ) external override returns (bytes4, BalanceDelta) {
+        PoolId poolId = key.toId();
+        bytes32 positionKey = CLPosition.calculatePositionKey(sender, params.tickLower, params.tickUpper, params.salt);
+
+        BalanceDelta hookDelta;
+        if (poolManager.getLiquidity(poolId) != 0) {
+            hookDelta = toBalanceDelta(
+                firstBlockFeesToken0[poolId][positionKey].toInt128(),
+                firstBlockFeesToken1[poolId][positionKey].toInt128()
+            );
+            poolManager.donate(
+                key, firstBlockFeesToken0[poolId][positionKey], firstBlockFeesToken1[poolId][positionKey], new bytes(0)
+            );
+        } else {
+            // If the pool is empty, the fees are not donated and are returned to the sender
+            hookDelta = BalanceDeltaLibrary.ZERO_DELTA;
+        }
+        return (this.afterRemoveLiquidity.selector, hookDelta);
```

Comment:
any reason why we are not doing the 'Cleanup stored data for the position'?

Reply:
same question here
maybe it's fine to not clean up if we are assuming that all users will only use position manager to add/removeLiq (position keys will always be different due to the fact that salt=tokenId)
but what if the user is not using position manager?
also as a plus, delete can save gas costs